### PR TITLE
HPE CSI Driver for Kubernetes v2.0.0

### PR DESCRIPTION
**Pull request by datamattsson**

### CODE REVIEW

File: charts/hpe-csi-driver/hpe-csi-driver/2.0.0/values.schema.json [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        "cspClientTimeout": {
+            "$id": "#/properties/cspClientTimeout",
+            "title": "CSP client timeout",
+            "description": "Change the CSP client timeout for HPE Alletra 9000, Primera and 3PAR systems",
+            "type": "integer",
+            "default": 60,
+            "minimum": 60,
+            "maximum": 360
+        }
+    },
```
Comment (PennyScissors) [inline charts/hpe-csi-driver/hpe-csi-driver/2.0.0/values.schema.json:167]: Please add `"global": {}` here to take care of the issue described in the other comment. Do not modify this file directly, instead make the change in this same file of the generate charts directory after running `make prepare` in order to save it in the patch.

**Review by PennyScissors — CHANGES_REQUESTED** [review]
Hello @datamattsson, I ran into issues deploying this chart in Rancher, but I had a hunch it was a problem on our side so I did some investigation. I found out we add values to charts during deployment that are needed for our system, which causes a validation error because your chart's json schema doesn't know about those values we add. Could you please patch in the change I requested in order to fix the issue? You will also need to add this same change to any new charts you may add in the future with a json schema to validate chart values.

File: charts/hpe-csi-driver/hpe-csi-driver/2.0.0/values.schema.json [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        },
+        "cspClientTimeout": {
+            "$id": "#/properties/cspClientTimeout",
+            "title": "CSP client timeout",
+            "description": "Change the CSP client timeout for HPE Alletra 9000, Primera and 3PAR systems",
+            "type": "integer",
+            "default": 60,
+            "minimum": 60,
+            "maximum": 360
+        }
```
Comment (PennyScissors) [inline charts/hpe-csi-driver/hpe-csi-driver/2.0.0/values.schema.json:166]: Please add the following to take care of the issue described in the other comment. Do not modify this file directly, instead make this change in the generated charts directory (after running `make prepare`) in order to save it in the patch.
```suggestion
        },
        "global": {}

**Review by PennyScissors — APPROVED** [review]
@datamattsson, I couldn't successfully deploy it in Rancher since I don't have a CSI and it tries to connect to one before all other resources, but this has been the same case with other HPE charts and I assume you've already tested it since it's part of the PR process so I'll go ahead and approve it.

**Comment by datamattsson** [comment]
The chart should install fine without any errors. This is on a Rancher 2.5.5 cluster:
```
helm install --namespace=hpe-storage --timeout=10m0s --values=/home/shell/helm/values-hpe-csi-driver-2.0.0.yaml --version=2.0.0 --wait=true hpe-csi-driver /home/shell/helm/hpe-csi-driver-2.0.0.tgz
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
creating 29 resource(s)
beginning wait for 29 resources with timeout of 10m0s
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 0 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 1 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 3 out of 4 expected pods are ready
DaemonSet is not ready: hpe-storage/hpe-csi-node. 3 out of 4 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
Deployment is not ready: hpe-storage/hpe-csi-controller. 0 out of 1 expected pods are ready
NAME: hpe-csi-driver
LAST DEPLOYED: Tue Jul 27 17:26:58 2021
NAMESPACE: hpe-storage
STATUS: deployed
REVISION: 1
TEST SUITE: None

---------------------------------------------------------------------
SUCCESS: helm install --namespace=hpe-storage --timeout=10m0s --values=/home/shell/helm/values-hpe-csi-driver-2.0.0.yaml --version=2.0.0 --wait=true hpe-csi-driver /home/shell/helm/hpe-csi-driver-2.0.0.tgz
---------------------------------------------------------------------
```

I also configured a backend secret and storageclass. Provisioning volumes and attaching workloads work just fine.

```
/tmp:mmattsson$ k create -f-
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-first-pvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 32Gi
persistentvolumeclaim/my-first-pvc created
(reverse-i-search)`':
/tmp:mmattsson$ k get pvc
NAME           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
my-first-pvc   Bound    pvc-4f1efa37-3c42-4fd1-a2cb-a9c2bdcb806e   32Gi       RWO            hpe-standard   4s
/tmp:mmattsson$ k create -f-
kind: Pod
apiVersion: v1
metadata:
  name: my-pod
spec:
  containers:
    - name: pod-datelog-1
      image: nginx
      command: ["bin/sh"]
      args: ["-c", "while true; do date >> /data/mydata.txt; sleep 1; done"]
      volumeMounts:
        - name: export1
          mountPath: /data
    - name: pod-datelog-2
      image: debian
      command: ["bin/sh"]
      args: ["-c", "while true; do date >> /data/mydata.txt; sleep 1; done"]
      volumeMounts:
        - name: export1
          mountPath: /data
  volumes:
    - name: export1
      persistentVolumeClaim:
        claimName: my-first-pvc
pod/my-pod created
/tmp:mmattsson$ k get pods -w
NAME     READY   STATUS              RESTARTS   AGE
my-pod   0/2     ContainerCreating   0          4s
my-pod   0/2     ContainerCreating   0          15s
my-pod   2/2     Running             0          38s
```

My newer Rancher clusters are behind a corporate proxy and can't reach github.com at the moment but I'm sure it's fine.

**Comment by PennyScissors** [comment]
Thanks for providing that extra information to see it running!

**Pull request merged**